cside-main/src/system_manager.c

```diff
@@ -1,14 +1,151 @@
+    ESP_ERROR_CHECK(ret);
+
+    int *context = (int *)malloc(sizeof(int));
+    *context = 0;
+}
+
+/* PUBLIC FUNCTIONS */
 system_manager_t* system_manager_init() {
-    return NULL;
+	assert(!_system_initialized && "System Manager was initialized a second time!");
+    system_manager_t* system_manager = (system_manager_t *) calloc(1, sizeof(system_manager_t));
+
+    system_manager->_connection_attempts = 0;
+    system_manager->wifi_connection = false;
+
+    _system_manager_board_init(system_manager);
+
+    _system_initialized = true;
+
+    return system_manager;
 }
 
-void system_manager_destroy(system_manager_t* sm) {
 
+void system_manager_destroy(system_manager_t* system_manager) {
+    if (system_manager == NULL) {
+        return;
+    }
+    // Free any allocated memory
+    if (system_manager->wifi_any_event_handle != NULL) {
+        free(system_manager->wifi_any_event_handle);
+    }
+    if (system_manager->got_ip_event_handle != NULL) {
+        free(system_manager->got_ip_event_handle);
+    }
+    free(system_manager);
 }
 
-bool system_manager_wifi_init(system_manager_t* sm) {
-    return NULL;
+
+void reset_system_manager(system_manager_t* system_manager) {
+    // Reset all fields to zero
+    memset(system_manager, 0, sizeof(system_manager_t));
+    // Reapply any specific default values
+    system_manager->_connection_attempts = 0;
+    system_manager->wifi_connection = false;
+
+    // Reset event handler pointers to NULL
+    system_manager->wifi_any_event_handle = NULL;
+    system_manager->got_ip_event_handle = NULL;
 }
+
+bool system_manager_wifi_init(system_manager_t* system_manager) {
+	s_event_group_handler = xEventGroupCreate();
+
+    ESP_ERROR_CHECK(esp_netif_init());
+
+    ESP_ERROR_CHECK(esp_event_loop_create_default());
+    esp_netif_create_default_wifi_sta();
+
+    wifi_init_config_t config = WIFI_INIT_CONFIG_DEFAULT();
+    ESP_ERROR_CHECK(esp_wifi_init(&config));
+
+    ESP_ERROR_CHECK(
+        esp_event_handler_instance_register(WIFI_EVENT, ESP_EVENT_ANY_ID, &event_handler, system_manager, &system_manager->wifi_any_event_handle));
+    ESP_ERROR_CHECK(
+        esp_event_handler_instance_register(IP_EVENT, IP_EVENT_STA_GOT_IP, &event_handler, system_manager, &system_manager->got_ip_event_handle));
+
+    wifi_config_t wifi_config = {.sta = {
+                                     .ssid = ESP_WIFI_SSID,
+                                     .password = ESP_WIFI_PASS,
+                                 }};
+
+    ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+    ESP_ERROR_CHECK(esp_wifi_set_config(WIFI_IF_STA, &wifi_config));
+    ESP_ERROR_CHECK(esp_wifi_start());
+
+    EventBits_t bits = xEventGroupWaitBits(s_event_group_handler, WIFI_CONNECTED_BIT, pdFALSE, pdFALSE, portMAX_DELAY);
+
+    if (bits & WIFI_CONNECTED_BIT) {
+        s_is_wifi_connected = true;
+    }
+    ESP_ERROR_CHECK(esp_event_handler_instance_unregister(IP_EVENT, IP_EVENT_STA_GOT_IP, system_manager->got_ip_event_handle));
```

Comment:
Here could set the wifi handles to NULL to avoid the assertion error when calling system_manager_destroy()

Reply:
is this something we necessary want to do? Are we certain that after connecting to the WIFI for the first time we want to remove the event handlers?

Reply:
From what I understand, lines 147-149 already delete (unregister) the wifi event handles ...
I am assuming that the memory allocated to these handles are freed inside of these functions, so we are able to use these event handlers anyways